### DISCUSSION TITLE
Using pb_hooks to send an email with attachments

### DISCUSSION THREAD

**Discussion by andrewpmoore** [pocketbase/pocketbase#6153] [Q&A]
I'm using javascript pocketbase hooks to send emails like this, so as a record get created into a table an email is sent and the table stamped with the `sentDate`.

```
onRecordCreate((e) => {
    const record = e.record;
    const toEmail = record.get("to");
    const subject = record.get("title");
    const body = record.get("body");


    if (!toEmail) {
        console.log("No 'to' destination provided.");
        return; // Stop processing the record
    }

    try {
        const message = new MailerMessage({
            from: {
                address: e.app.settings().meta.senderAddress,
                name: e.app.settings().meta.senderName,
            },
            to: [{address: toEmail}],
            subject: subject,
            html: body,


        });

        e.app.newMailClient().send(message)
        record.set("sentDate", new Date().toISOString());
    } catch (err) {
        console.error("Error during sending email:", err.message);
    }

    e.next()
}, "emails")
```

I've now added a `pdf` field which is defined as a `file`. I'm just not sure what I need to do to add the file to the email sent. I've tried adding:

`    const pdf = record.get("pdf");  `
    
 and then passing in 
`      attachments: [{pdf}]`
      
to the mailer, but that doesn't lead to an attachment on the email. 
Any ideas what I need to do to attach the file?

**Comment by ganigeorgiev** [comment]
The attachments must be a "name-io.Reader" pair. _The file is not uploaded yet, so the returned record value is of [`filesystem.File` instance](https://pocketbase.io/jsvm/interfaces/filesystem.File.html) and you can get a io.Reader from it by calling `yourFile.reader.open()`._

Your example could look something like:

```js
onRecordCreateExecute((e) => {
    const record = e.record;
    const toEmail = record.get("to"); // note: if "to" is required mark the field as "Nonempty"
    const subject = record.get("title");
    const body = record.get("body");
    const pdf = record.get("pdf");

    let file;

    try {
        const message = new MailerMessage({
            from: {
                address: e.app.settings().meta.senderAddress,
                name: e.app.settings().meta.senderName,
            },
            to: [{address: toEmail}],
            subject: subject,
            html: body,
            // blank map is required for now because otherwise it will be undefined and will fail if we try to populate it directly from outside
            // (I'll consider initialising it with the next release)
            attachments: {},
        });

        // has file
        if (pdf) {
            file = pdf.reader.open();
            message.attachments[pdf.name] = file
        }

        e.app.newMailClient().send(message)
        record.set("sentDate", new Date().toISOString());
    } catch (err) {
        // note: you can also use the builtin logger if you want to be stored in the database
        // https://pocketbase.io/docs/js-logging/
        // (logs are printed by default in the console when running with "--dev")

        console.error("Error during sending email:", err.message);
    } finally {
        file?.close();
    }

    e.next()
}, "emails")
```

_edit: replaced the hook with [`onRecordCreateExecute`](https://pocketbase.io/docs/js-event-hooks/#onrecordcreateexecute) so that it runs after the Record validations._
_edit2: updated the example to use uploaded file instance._

**Comment by andrewpmoore** [comment]
Thanks for your prompt reply as usual @ganigeorgiev 
That's helped me understand it a little more. I'm getting `Error during sending email: blob not found`
I think it comes from `const pdfName = record.get("pdf");` 
...as if I print out, it shows as `[object Object]`  but then it's used in getting the file
`file = fsys.getFile(record.baseFilesPath() + "/" + pdfName)`

I guess there's a property of the object I need to get to.
Apologies, I'm more of a dart/flutter person and I struggle with javascript!

**Reply by ganigeorgiev** [reply]
I assume that you use PocketBase v0.23+, right?

~In that case I'm not able to reproduce it. Calling `console.log(record.get("pdf"))` should return string, unless `pdf` is not a single `file` field. Could you export your collections configuration as json (Dashboard > Settings > Export collections)?~

_I've updated the example._

**Comment by andrewpmoore** [comment]
Yeah, I'm on v0.23
Here's the export
[export.json](https://github.com/user-attachments/files/18213546/export.json)

**Reply by ganigeorgiev** [reply]
Sorry my mistake, I was testing with the `onRecordAfterCreateSuccess` hook.

In your case it returns an object because the record file value is replaced with the resulting filename only after successful `e.next()`, aka. it is technically not considered uploaded yet and the value is instance of [`filesystem.File`](https://pocketbase.io/jsvm/interfaces/filesystem.File.html) to allow developers to modify the file before persistence or to add a new one programmatically via the [$filesystem.*](https://pocketbase.io/jsvm/modules/_filesystem.html) helpers.

I've updated the example with the correct code and removed the older comments to prevent confusion.

**Comment by andrewpmoore** [comment]
Perfect, that's fixed it, thanks again for pocketbase and the excellent support you provide 👍
